bittensor/_axon/__init__.py

```diff
@@ -484,13 +498,19 @@ def intercept_service(self, continuation, handler_call_details):
                 signature_format,
             ) = self.parse_signature(metadata)
 
+            # Build the key which uniquely identifies the endpoint that has signed the message.
+            endpoint_key = f"{sender_hotkey}:{receptor_uuid}"
+            
             # signature checking
-            self.check_signature(
-                nonce, sender_hotkey, signature, receptor_uuid, signature_format
-            )
+            self.check_signature(nonce, sender_hotkey, signature, receptor_uuid, signature_format, endpoint_key)
 
             # blacklist checking
             self.black_list_checking(sender_hotkey, method)
+            
+            # update nonce when all checking pass
```

Comment:
If you write the nonce only after the blacklist check, the nonce may be reused if the blacklist check fails.